### PR TITLE
AWS: Enable clusters with no public endpoints

### PR DESCRIPTION
**Pull request by abhinavdahiya**
pkg/types: add the publish strategy to install-config

This commit adds a enum(string) type PublishingStrategy. It supports 2 options

* ExternalPublishingStrategy : the endpoints are exposed to the Internet
* InternalPublishingStrategy : the endpoints are exposed to the `Private Network` only

This enum type is added to the InstallConfig to control the strategy for the cluster endpoints. Cluster endpoints include the API server,
the default Ingress controller, public IPs etc.

Also adds the defaulting to make sure this strategy is `External` by default, keeping the backward-compatible behavior, and validation to make sure only valid enum options are set.

 pkg/asset: update machines and DNS for external-internal strategy

For aws, only when the strategy is External, do we set the public Route53 zone for the basedomain in DNSes.config.openshift.com cluster object.

 data/aws: add publish strategy External/Internal feature

Adds the `aws_publish_strategy` variable to the terraform root variables to allow skipping public resources.

The brief list of resources that are only created when strategy is External are:

- Public Load Balancers for API (6443) (aws_lb.api_external, aws_lb_target_group.api_external, aws_lb_listener.api_external_api)
- Public DNS record (aws_route53_record.api_external)
- No public IP address associated to bootstrap-host (associate_public_ip_address: false)
- No security group rule that allows SSH to bootstrap-host from Internet (aws_security_group_rule.ssh), rather switching to VPC only.

Due to terraform issue hashicorp/terraform#12570, where count cannot use variables that are dynamic as they are not computed during plan, certain implicit assumes/hacks need to be continued

An error below shows how the aws_lb_target list can't be used as sole source for allowing other modules like bootstrap and master attach isntances without caring about publish strategy.
Similarly the error also shows how route53 module cannot turn off the public records based on existence of the external LBs alone.

```
ERROR Error: Invalid count argument
ERROR
ERROR   on ../../../../../../../tmp/openshift-install-940637305/bootstrap/main.tf line 154, in resource "aws_lb_target_group_attachment" "bootstrap":
ERROR  154:   count = length(var.target_group_arns)
ERROR
ERROR The "count" value depends on resource attributes that cannot be determined
ERROR until apply, so Terraform cannot predict how many instances will be created.
ERROR To work around this, use the -target argument to first apply only the
ERROR resources that the count depends on.
ERROR
ERROR
ERROR Error: Invalid count argument
ERROR
ERROR   on ../../../../../../../tmp/openshift-install-940637305/master/main.tf line 131, in resource "aws_lb_target_group_attachment" "master":
ERROR  131:   count = var.instance_count * length(var.target_group_arns)
ERROR
ERROR The "count" value depends on resource attributes that cannot be determined
ERROR until apply, so Terraform cannot predict how many instances will be created.
ERROR To work around this, use the -target argument to first apply only the
ERROR resources that the count depends on.
ERROR
ERROR
ERROR Error: Invalid count argument
ERROR
ERROR   on ../../../../../../../tmp/openshift-install-940637305/route53/base.tf line 2, in data "aws_route53_zone" "public":
ERROR    2:   count = var.api_external_lb_dns_name != null ? 1 : 0
ERROR
ERROR The "count" value depends on resource attributes that cannot be determined
ERROR until apply, so Terraform cannot predict how many instances will be created.
ERROR To work around this, use the -target argument to first apply only the
ERROR resources that the count depends on.
```

For instances to communicate with internet, they need to
- either have public IP in a public subnet or,
- be launched in private subnet.

For more info see https://forums.aws.amazon.com/thread.jspa?threadID=96369

Therefore the bootstrap instance is created in public subnet for External strategy and private subnet for others

Also the bootstrap instance SSH rules are modified to target Internet (0.0.0.0) for External and allow only VPC CIDR ranges otherwise

xref: https://jira.coreos.com/browse/CORS-1221

/cc @wking

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
Create your VPC+subnets
```
$ aws cloudformation create-stack  --stack-name adahiya-byo-vpc --template-body "$(cat upi/aws/cloudformation/01_vpc.yaml)" --parameters ParameterKey=AvailabilityZoneCount,ParameterValue=3
$ aws cloudformation describe-stacks --stack-name adahiya-byo-vpc --query 'Stacks[].Outputs[]' --output json
```

Create bastion into your VPC.
````
$ cat upi/aws/cloudformation/01b_bastian.yaml
AWSTemplateFormatVersion: 2010-09-09
Description: Template for OpenShift Cluster Security Elements (Security Groups & IAM)

Parameters:
  VpcId:
    Description: The VPC-scoped resources will belong to this VPC.
    Type: AWS::EC2::VPC::Id
  SubnetId:
    Description: The public subnet.
    Type: AWS::EC2::Subnet::Id
  AllowedSshCidr:
    AllowedPattern: ^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(\/([0-9]|1[0-9]|2[0-9]|3[0-2]))$
    ConstraintDescription: CIDR block parameter must be in the form x.x.x.x/0-32.
    Default: 0.0.0.0/0
    Description: CIDR block to allow SSH access to the bastian host.
    Type: String
  Ami:
    Description: The AMI to use for bastian.
    Type: AWS::EC2::Image::Id
  SshKeyName:
    Description: The key pair in AWS for bastian.
    Type: String

Metadata:
  AWS::CloudFormation::Interface:
    ParameterGroups:
    - Label:
        default: "Network Configuration"
      Parameters:
      - VpcId
      - SubnetId
      - AllowedSshCidr
    - Label:
        default: "Host Information"
      Parameters:
      - Ami
      - SshKeyName
    ParameterLabels:
      VpcId:
        default: "VPC ID"
      SubnetId:
        default: "Public Subnet"
      AllowedSshCidr:
        default: "Allowed SSH Source"
      Ami:
        default: "AMI ID"
      Ami:
        default: "The key pair name"

Resources:
  BastianSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: Bastian Security Group
      SecurityGroupIngress:
      - IpProtocol: tcp
        FromPort: 22
        ToPort: 22
        CidrIp: !Ref AllowedSshCidr
      VpcId: !Ref VpcId

  BastianInstance:
    Type: AWS::EC2::Instance
    Properties:
      ImageId: !Ref Ami
      InstanceType: "m5.large"
      NetworkInterfaces:
      - AssociatePublicIpAddress: "true"
        DeviceIndex: "0"
        GroupSet:
        - !Ref BastianSecurityGroup
        SubnetId: !Ref SubnetId
      KeyName: !Ref SshKeyName

Outputs:
  BastianInstanceId:
    Description: Bastian Instance ID.
    Value: !Ref BastianInstance

  BastianPublicIp:
    Description: The bootstrap node public IP address.
    Value: !GetAtt BastianInstance.PublicIp


$ aws cloudformation create-stack  --stack-name adahiya-byo-vpc-bastian-2 --template-body "$(cat upi/aws/cloudformation/01b_bastian.yaml)" --parameters ParameterKey=VpcId,ParameterValue=vpc-0f582a396680d7058 ParameterKey=SubnetId,ParameterValue=subnet-00d14380eeb2b1111 ParameterKey=Ami,ParameterValue=ami-0e05097610fae5559 ParameterKey=SshKeyName,ParameterValue=adahiya
$ aws cloudformation describe-stacks --stack-name adahiya-byo-vpc-bastian-2 --query 'Stacks[].Outputs[]' --output json
```

I'm using sshuttle to create a tunnel to the VPC using ssh, it also handles DNS resolution
```
$ sshuttle --dns -r ec2-user@34.201.100.240 0/0
[local sudo] Password:
client: Connected.
```

Open another terminal and run the installer with your list of subnets and publish: Internal
```
$ openshift-install create cluster --dir dev
INFO Consuming Install Config from target directory
INFO Creating infrastructure resources...
INFO Waiting up to 30m0s for the Kubernetes API at https://api.adahiya-1.devcluster.openshift.com:6443...
INFO API v1.16.0-beta.2+a8701aa up
INFO Waiting up to 30m0s for bootstrapping to complete...
INFO Waiting up to 30m0s for the cluster at https://api.adahiya-1.devcluster.openshift.com:6443 to initialize...
INFO Waiting up to 10m0s for the openshift-console route to be created...
INFO Install complete!
INFO To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/home/adahiya/go/src/github.com/openshift/installer/dev/auth/kubeconfig'
INFO Access the OpenShift web-console here: https://console-openshift-console.apps.adahiya-1.devcluster.openshift.com
INFO Login to the console with user: kubeadmin, password: xxxx-xxxxx-xxxxx-xxxxx
```

**Comment by wking** [comment]
> I'm using sshuttle to create a tunnel to the VPC using ssh...

Could you SSH in and run the installer on the bastion?  I'd expect you'd have AWS-endpoint access to any cluster-managed AWS resources you'd need, because the cluster would be managing those resources going forward.  And it would get us one step closer to the installer-appliance AMI flow.

File: pkg/types/installconfig.go [hunk 1]
```diff
@@ -90,6 +100,11 @@ type InstallConfig struct {
 	// ImageContentSources lists sources/repositories for the release-image content.
 	// +optional
 	ImageContentSources []ImageContentSource `json:"imageContentSources,omitempty"`
+
+	// Publish controls how the user facing endpoints of the cluster like the k8s API, OpenShift routes etc. are exposed.
```
Comment (wking) [inline pkg/types/installconfig.go]: nit: can we spell out `k8s` -> `Kubernetes` in docs like this?

File: pkg/types/validation/installconfig.go [hunk 2]
```diff
@@ -304,3 +307,18 @@ func validateNamedRepository(r string) error {
 	}
 	return nil
 }
+
+var (
+	validPublishingStrategies = map[types.PublishingStrategy]bool{
+		types.ExternalPublishingStrategy: true,
```
Comment (wking) [inline pkg/types/validation/installconfig.go]: nit: in other parts of this repo, we use `map[...]struct{}` for existence-checking sets.  Dunno if we care about consistency; I'm fine either way.

File: pkg/types/validation/installconfig_test.go [hunk 3]
```diff
@@ -700,6 +701,15 @@ func TestValidateInstallConfig(t *testing.T) {
 				return c
 			}(),
 		},
+		{
+			name: "invalid publishing strategy",
+			installConfig: func() *types.InstallConfig {
+				c := validInstallConfig()
+				c.Publish = types.PublishingStrategy("ExternalInternalDoNotCare")
```
Comment (wking) [inline pkg/types/validation/installconfig_test.go]: :joy:

File: pkg/types/validation/installconfig.go [hunk 4]
```diff
@@ -304,3 +307,18 @@ func validateNamedRepository(r string) error {
...
+		types.InternalPublishingStrategy: true,
+	}
+
+	validPublishingStrategyValues = func() []string {
+		v := make([]string, 0, len(validPublishingStrategies))
+		for m := range validPublishingStrategies {
+			v = append(v, string(m))
+		}
+		return v
```
Comment (wking) [inline pkg/types/validation/installconfig.go]: nit: sort the slice before returning?

File: pkg/tfvars/aws/aws.go [hunk 5]
```diff
@@ -23,11 +25,12 @@ type config struct {
 	Region                  string            `json:"aws_region,omitempty"`
 	VPC                     string            `json:"aws_vpc,omitempty"`
 	PrivateSubnets          []string          `json:"aws_private_subnets,omitempty"`
-	PublicSubnets           []string          `json:"aws_public_subnets,omitempty"`
+	PublicSubnets           *[]string         `json:"aws_public_subnets,omitempty"`
```
Comment (wking) [inline pkg/tfvars/aws/aws.go]: I don't think we need this pointer: https://play.golang.org/p/yf1LDeu2r1M

Comment (wking) [reply pkg/tfvars/aws/aws.go]: Oh, I see.  You want the explicit `[]` option.  All good with the pointer then.

File: data/data/aws/vpc/variables.tf [hunk 6]
```diff
@@ -11,14 +11,10 @@ variable "cluster_id" {
...
-  description = "If set to true, private-facing ingress resources are created."
-  default     = true
-}
-
-variable "public_master_endpoints" {
-  description = "If set to true, public-facing ingress resources are created."
-  default     = true
+variable "publish_strategy" {
+  type        = string
+  default     = "External"
```
Comment (wking) [inline data/data/aws/vpc/variables.tf]: I'd rather not have a default, to ensure we are successfully passing this in from Go.

File: data/data/aws/vpc/variables.tf [hunk 7]
```diff
@@ -11,14 +11,10 @@ variable "cluster_id" {
...
+  description = "The publishing strategy for endpoints like Load balancers"
```
Comment (wking) [inline data/data/aws/vpc/variables.tf]: nit: "Load" -> "load".  Or, if you go AWS-proper-name, "Load Balancers".

File: pkg/tfvars/aws/aws.go [hunk 8]
```diff
@@ -85,7 +88,15 @@ func TFVars(vpc string, privateSubnets []string, publicSubnets []string, masterC
 		Type:                    *rootVolume.EBS.VolumeType,
 		VPC:                     vpc,
 		PrivateSubnets:          privateSubnets,
-		PublicSubnets:           publicSubnets,
+		PublishStrategy:         string(publish),
+	}
+
+	if len(publicSubnets) == 0 {
+		if cfg.VPC != "" {
+			cfg.PublicSubnets = &[]string{}
```
Comment (wking) [inline pkg/tfvars/aws/aws.go]: I feel like we should support installer-provided VPCs with internal publishing, where the user expects to create a VPN into the cluster before the waiting-for-Kube-API bit times out.  In that case, we'd want to switch on `publish` here.  But that would definitely be a corner case.  Thoughts?

**Comment by abhinavdahiya** [comment]
> > I'm using sshuttle to create a tunnel to the VPC using ssh...
> 
> Could you SSH in and run the installer on the bastion?

It would mean I have to ssh, and move all the binaries, configuration, secrets to the Bastian and then run the installer, sshuttle was extremely useful for me to develop and iterate easily. Bonus sshuttle captured my DNS and I could access the cluster console through my browser... 

Most of our customers will have their computers already configured to have access to network only private resources.. it was easy for me...

> I'd expect you'd have AWS-endpoint access to any cluster-managed AWS resources you'd need, because the cluster would be managing those resources going forward. And it would get us one step closer to the installer-appliance AMI flow.

I agree 💯 but a UI based installer is very important to make that easy to use.

File: pkg/tfvars/aws/aws.go (hunk 8 above)
Comment (abhinavdahiya) [reply pkg/tfvars/aws/aws.go]: > I feel like we should support installer-provided VPCs with internal publishing,

Well that would be possible in this setup, as in that case public subnets will be set as nil ie create for me and vpc is also empty....

> Thoughts?

Currently the validation is not in place, and there will be following up PRs for that, and create vpc + internal will be explicitly not allowed as we specified in the design. The reason is the UX for ipi in this case is going to be too over the place that we don't want to support it right now.

File: pkg/types/installconfig.go (hunk 1 above)
Comment (abhinavdahiya) [reply pkg/types/installconfig.go]: fixed with https://github.com/openshift/installer/compare/586800a7998182fd87bc00700adf24a49df216be...83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d

File: pkg/types/validation/installconfig.go (hunk 2 above)
Comment (abhinavdahiya) [reply pkg/types/validation/installconfig.go]: fixed with https://github.com/openshift/installer/compare/586800a7998182fd87bc00700adf24a49df216be...83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d

File: data/data/aws/vpc/variables.tf (hunk 6 above)
Comment (abhinavdahiya) [reply data/data/aws/vpc/variables.tf]: fixed with https://github.com/openshift/installer/compare/586800a7998182fd87bc00700adf24a49df216be...83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d

File: data/data/aws/vpc/variables.tf (hunk 7 above)
Comment (abhinavdahiya) [reply data/data/aws/vpc/variables.tf]: fixed with https://github.com/openshift/installer/compare/586800a7998182fd87bc00700adf24a49df216be...83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d

File: pkg/types/validation/installconfig.go (hunk 4 above)
Comment (abhinavdahiya) [reply pkg/types/validation/installconfig.go]: fixed with https://github.com/openshift/installer/compare/586800a7998182fd87bc00700adf24a49df216be...83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d

**Comment by abhinavdahiya** [comment]
All green :)

File: data/data/aws/vpc/variables.tf (hunk 6 above)
Comment (wking) [reply data/data/aws/vpc/variables.tf]: still have a default over [here][1].

[1]: https://github.com/openshift/installer/blob/83adc04d6cf70ffc2f45bbe6a5e8d1d06173832d/data/data/aws/master/variables.tf#L75

File: data/data/aws/master/main.tf [hunk 9]
```diff
@@ -1,5 +1,9 @@
 locals {
   arn = "aws"
+
+  // Because of the issue https://github.com/hashicorp/terraform/issues/12570, the consumers cannot use a dynamic list for count
+  // and therefore are force to implicitly assume that the list is of aws_lb_target_group_arns_length - 1, in case there is no api_external
```
Comment (wking) [inline data/data/aws/master/main.tf]: nit: 12570 was [closed with 0.12's better error message][1], and the actual support tracking has moved to [terraform#4149][2].  Dunno if it's worth updating your links?

[1]: https://github.com/hashicorp/terraform/issues/12570#issuecomment-512621787
[2]: https://github.com/hashicorp/terraform/issues/4149

File: data/data/aws/route53/variables.tf [hunk 10]
```diff
@@ -54,3 +54,15 @@ variable "api_internal_lb_zone_id" {
...
+variable "publish_strategy" {
+  type        = string
+  default     = "External"
+  description = <<EOF
+The publishing strategy for endpoints like load balancers
+
+Because of the issue https://github.com/hashicorp/terraform/issues/12570, the consumers cannot count 0/1
+based on if api_external_lb_dns_name for example, which will be null when there is no external lb for API.
+So publish_strategy serves an coordinated proxy for that decision.
+
```
Comment (wking) [inline data/data/aws/route53/variables.tf]: nit: empty line before the EOF

**Comment by wking** [comment]
[Left][1] [three][2] [nits][3], but I don't care if they get fixed in this PR, in follow-up work, or never.  Pull the hold if you want to punt on them.

/lgtm
/hold

[1]: https://github.com/openshift/installer/pull/2526#discussion_r336697111
[2]: https://github.com/openshift/installer/pull/2526#discussion_r336699055
[3]: https://github.com/openshift/installer/pull/2526#discussion_r336694928

File: data/data/aws/route53/variables.tf (hunk 10 above)
Comment (abhinavdahiya) [reply data/data/aws/route53/variables.tf]: fixed: https://github.com/openshift/installer/compare/09284a49abb5433054f290c35e0aa14a98eef974...2bd7e958c24af56fb51e653786ea3ad37f6d25f4

File: data/data/aws/vpc/variables.tf (hunk 6 above)
Comment (abhinavdahiya) [reply data/data/aws/vpc/variables.tf]: fixed https://github.com/openshift/installer/compare/09284a49abb5433054f290c35e0aa14a98eef974...2bd7e958c24af56fb51e653786ea3ad37f6d25f4

File: data/data/aws/master/main.tf (hunk 9 above)
Comment (abhinavdahiya) [reply data/data/aws/master/main.tf]: I think 12570 definitely has the correct context

**Comment by abhinavdahiya** [comment]
/hold cancel

fixed [two][1] [nits][2] from https://github.com/openshift/installer/pull/2526#issuecomment-543988578

[1]: https://github.com/openshift/installer/pull/2526#discussion_r336699055
[2]: https://github.com/openshift/installer/pull/2526#discussion_r336694928

**Pull request merged**